src/meta/src/barrier/schedule.rs

```diff
@@ -80,13 +79,27 @@ impl<S: MetaStore> BarrierScheduler<S> {
 
     /// Attach `new_notifiers` to the very first scheduled barrier. If there's no one scheduled, a
     /// default checkpoint barrier will be created.
-    async fn attach_notifiers(&self, new_notifiers: impl IntoIterator<Item = Notifier>) {
+    async fn attach_notifiers(&self, new_notifiers: Vec<Notifier>) {
         let mut queue = self.inner.queue.write().await;
+        let new_checkpoint = new_notifiers
+            .iter()
+            .any(|notifier| notifier.bound_to_checkpoint_barrier());
```

Comment:
I think it's kind of weird to add a flag of `checkpoint` in the notifier. For now, the `checkpoint` flags in the notifiers seem redundant to `Scheduled::checkpoint`. To tell that we should flush with checkpoints, I think we may simple add a flag to the parameters of this function, and there's no need to store it anywhere.